### DISCUSSION TITLE
How to write a "BaseRouter" to reduce redundant code?

### DISCUSSION THREAD

**Discussion by CheelHorn** [fastapi/fastapi#6113] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.get("/{exercise_id}", response_model=Exercise)
async def get_exercise(
    exercise_id: UUID4,
    exercise_service: ExcerciseService = Depends(get_exercise_service),
) -> Optional[models.Exercise]:
    return exercise_service.get(exercise_id)

@router.get("/{user_id}", response_model=User)
async def get_user(
    user_id: UUID4,
    user_service: UsersService = Depends(get_users_service),
) -> Optional[models.User]:
    return user_service.get(user_id)
```

### Description

I have several models which all require the basic CRUD operations as API endpoints. Therefore I wrote several routers containing similar code for each route (get, list, create, update, delete).  In the code example I included two routes getting an object by its ID. The only difference is the request and response model and the called backend function to handle the requested data. I was wondering if there is any way to create a BaseRouter which the other Routers can inherit from and which contains these basic routes that can then be adjusted if needed. I think I am looking for something like class based views, but I am not sure if that is the correct terminology

I opened a [Gituhub issue](https://stackoverflow.com/questions/75249150/how-to-use-class-based-views-in-fastapi) with a simplified example but the answers did not lead to a satisfactory answer.
I looked at [FastApi Utils](https://fastapi-utils.davidmontague.xyz/) but this did not seem to help either.

Especially regarding the Dependency Injection I did not find a solution.

### Operating System

Windows

### FastAPI Version

0.89.1

### Python Version

3.10.6

**Comment by ebottos94** [comment]
@CheelHorn if the example is referred to the same router the second route will be never called beacuse the path definition will be the same, if are referred to two different router is ok. Regarding your question, if your crud operations are similar in every routers you can declare a generic function that execute your statement passed as a parameter. For example in your case with SQLModel you can do something like this :

```python
@router.get("/{exercise_id}", response_model=Exercise)
async def get_exercise(
    exercise_id: UUID4
) -> Optional[models.Exercise]:
    statement = select(Exercise).where(Exercise.exercise_id == exercise_id)
    return get_service(statement)

@router.get("/{user_id}", response_model=User)
async def get_user(
    user_id: UUID4
) -> Optional[models.User]:
    statement = select(User).where(User.user_id == user_id)
    return get_service(statement)

def get_service(statement):
    with Session(engine) as session:
        result = session.exec(statement)
        ...
```

I hope I was helpful

**Comment by CheelHorn** [comment]
@ebottos94 Thanks for your response. The routes are of course for different routers, I just wanted to show the similarities between the two routes. I split up my project into different router files.

My question was not regarding the function call inside the route but the route itself. For my model Exercise and my model User I need 5 basic routes (get, list, create, update, delete). In my example above is the get route. These routes look the same for every Model, the only differences are the request and response models and the services they are depending on (exercise_service and user_service). If I add more models I need to over and over write the same routes again. Therefore I was wondering if it is possible to create a BaseRouter which defines these basic routes (get, list, create, update, delete) and the Exercise route for example inherites from that router and id necessary some routes can be overwritten. This would make the code more maintainable and future changes easier.

I hope my problem and question is more clear now.

**Comment by ebottos94** [comment]
@CheelHorn it's possible I'm wrong but I don't think there's a way to do this in fastapi.

**Comment by yinziyan1206** [comment]
I guess what you need is a code generator. You can just use `jinja` (or etc) to create a template

**Comment by CheelHorn** [comment]
Hmmm alright, thanks for the answers though.

**Comment by code404sg** [comment]
i use this as base.
``` python
class BaseAPI:
    session=Session(engine)
    model_name = Role
    # model_name = BaseModel
    route_name = "/base"
    tags = ["default"]
    def __init__(self):
        print(self.__class__.__qualname__)
        self.router = APIRouter(prefix=f"{self.route_name}", tags=self.tags)
        self.fetch_api_route()
        self.create_api_route()
        # self.router.add_api_route('/{id}', method=["GET"])
        # self.router.add_api_route('/{id}', method=["PUT"])
    def fetch_api_route(self):
        self.router.add_api_route('/', methods=["GET"], endpoint=self.fetch_all, response_model=None)
    def fetch_all(self, q:Optional[str]=None, page:Optional[int]=1, limit:Optional[int]=100):
        statement = select(self.model_read).limit(limit).offset(page)
        results = self.session.exec(statement)
        return results.all()
    
    def create_api_route(self):
        self.router.add_api_route('/', methods=["POST"], endpoint=self.create, response_model=None)
    def create(self, data: dict):
        # self.model_name(*data)
        print("checking validate")
        print(self.model_create.validate(data))
        record = self.model_read(**data)
        self.session.add(record)
        self.session.commit()
        self.session.close()
        return record
```

and this is the class

``` python
class RoleAPI(BaseAPI):
    model_read = Role
    model_create = RoleCreate
    route_name = "/roles"
    tags = ["roles"]
```

i dont know if this is the correct way.
